### DISCUSSION TITLE
Accessing Fabric functionality via REST APIs

### DISCUSSION THREAD

**Discussion by LeonidShamis** [danielmiessler/Fabric#1114] [General]
Hi,

I was wondering if it is possible to use Fabric via REST APIs to achieve the same functionality that is available with the CLI client.

I see that "--serve" option starts the Fabric Rest API server and the patterns, sessions and contexts endpoints become available as shown below.

However, I couldn't work out how to perform actions using patterns, e.g. how to submit a text for the "summarize" pattern, how to submit a YouTube URL for the "extract_wisdom", etc.

Essentially, I'm looking for a way to use full Fabric functionality through REST API.

Thank you.

```
[GIN-debug] GET    /patterns/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func1 (5 handlers)
[GIN-debug] GET    /patterns/names           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func2 (5 handlers)
[GIN-debug] DELETE /patterns/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func3 (5 handlers)
[GIN-debug] GET    /patterns/exists/:name    --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func4 (5 handlers)
[GIN-debug] PUT    /patterns/rename/:oldName/:newName --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func5 (5 handlers)
[GIN-debug] POST   /patterns/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func6 (5 handlers)
[GIN-debug] GET    /contexts/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func1 (5 handlers)
[GIN-debug] GET    /contexts/names           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func2 (5 handlers)
[GIN-debug] DELETE /contexts/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func3 (5 handlers)
[GIN-debug] GET    /contexts/exists/:name    --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func4 (5 handlers)
[GIN-debug] PUT    /contexts/rename/:oldName/:newName --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func5 (5 handlers)
[GIN-debug] POST   /contexts/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func6 (5 handlers)
[GIN-debug] GET    /sessions/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func1 (5 handlers)
[GIN-debug] GET    /sessions/names           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func2 (5 handlers)
[GIN-debug] DELETE /sessions/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func3 (5 handlers)
[GIN-debug] GET    /sessions/exists/:name    --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func4 (5 handlers)
[GIN-debug] PUT    /sessions/rename/:oldName/:newName --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func5 (5 handlers)
[GIN-debug] POST   /sessions/:name           --> github.com/danielmiessler/fabric/restapi.NewStorageHandler[...].func6 (5 handlers)
```

**Comment by noamsiegel** [comment]
As far as I know, there is no REST API for fabric. 

It would be difficult to do so since receiving a response requires using an API key. However, passing them through to any fabric function call would be possible. 

If you are using it locally, you could run the commands using a terminal session.

**Reply by LeonidShamis** [reply]
Hi @noamsiegel, thank you for your response.

The scenario that I'm interested in is using Fabric functionality remotely through REST APIs. I can use it from a command line when Fabric is installed locally and I already configured it with a local or remote LLM provider. It would be terrific to have access to the same functionality from a remote using REST APIs.

When you are talking about passing an API key, I guess you are referring to a way of securing Fabric (service) for users' access using an API key. I think that it could be achieved using an API Gateway in front of Fabric, taking responsibility for securing REST API endpoints, managing users and API keys, and managing authentication.

The missing piece remains the lack of all REST API endpoints for accessing full Fabric functionality when starting the Fabric with a "--serve" option, which starts the Fabric Rest API server, but currently has only the patterns, sessions and contexts endpoints as shown above.

Kind Regards,
Leonid

**Reply by alexissinglaire** [reply]
Hi @LeonidShamis ,

You can use following sample of JSON request message using POSTMAN to send to Fabric REST using ollama format. Hope it helps. 

_PS: I'm using Ollama endpoint as the default Vendor ._

****to activate REST api to use Ollama message format, you need to change the Dockerfile to following**
CMD ["--serveOllama"]** 

HTTP method: POST
HTTP url: http://<IP address>:<port>/api/chat
HTTP header: 
  Content-Type: application/json

HTTP Body:

{
  "model": "ask_uncle_duke",

  "stream": false, 
  "messages": [
    { 
      "role": "user",
      "content": "why is the space dark?"
    }
  ]
}
